src/Symfony/Component/Validator/Tests/Constraints/ExpressionLanguageProviderTest.php

```diff
@@ -0,0 +1,117 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Constraints;
+
+use PHPUnit\Framework\TestCase;
+use Symfony\Component\ExpressionLanguage\ExpressionLanguage;
+use Symfony\Component\Validator\Constraints\ExpressionLanguageProvider;
+use Symfony\Component\Validator\Constraints\NotNull;
+use Symfony\Component\Validator\Constraints\Range;
+use Symfony\Component\Validator\ConstraintViolationListInterface;
+use Symfony\Component\Validator\Context\ExecutionContextInterface;
+use Symfony\Component\Validator\Validator\ContextualValidatorInterface;
+use Symfony\Component\Validator\Validator\ValidatorInterface;
+
+class ExpressionLanguageProviderTest extends TestCase
+{
+    /**
+     * @dataProvider dataProviderCompile
+     */
+    public function testCompile(string $expression, array $names, string $expected): void
+    {
+        $provider = new ExpressionLanguageProvider();
+
+        $expressionLanguage = new ExpressionLanguage();
+        $expressionLanguage->registerProvider($provider);
+
+        $result = $expressionLanguage->compile($expression, $names);
+
+        $this->assertSame($expected, $result);
+    }
+
+    public function dataProviderCompile(): array
+    {
+        return [
+            [
+                'is_valid("foo", constraints)',
+                ['constraints'],
+                '0 === $context->getValidator()->inContext($context)->validate("foo", $constraints)->getViolations()->count()',
+            ],
+            [
+                'is_valid(this.data, constraints, groups)',
+                ['this', 'constraints', 'groups'],
+                '0 === $context->getValidator()->inContext($context)->validate($this->data, $constraints, $groups)->getViolations()->count()',
+            ]
+        ];
+    }
+
+    /**
+     * @dataProvider dataProviderEvaluate
+     */
+    public function testEvaluate(bool $expected, int $errorsCount): void
+    {
+        $constraints = [new NotNull(), new Range(['min' => 2])];
+
+        $violationList = $this->getMockBuilder(ConstraintViolationListInterface::class)
+            ->onlyMethods(['count'])
+            ->getMockForAbstractClass();
```

Comment:
why all this chain? wouldn't `createMock()` be enough? (same question elsewhere)

Reply:
By the way: Can we find a way to not use any mocks here at all?

Reply:
Changed